pkg/queue/request_log.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"text/template"
+	"time"
+
+	pkghttp "github.com/knative/serving/pkg/http"
+	"github.com/knative/serving/pkg/utils"
+)
+
+type requestLogHandler struct {
+	handler  http.Handler
+	writer   io.Writer
+	template *template.Template
+}
+
+// NewRequestLogHandler creates an http.Handler that logs request logs to an io.Writer.
+func NewRequestLogHandler(h http.Handler, w io.Writer, templateStr string) (http.Handler, error) {
+	// Make sure that the template ends with a newline. Otherwise,
+	// logging backends will not be able to parse entries separately.
+	if !strings.HasSuffix(templateStr, "\n") {
+		templateStr = templateStr + "\n"
+	}
+
+	// Expose a function to give access to cached environment variables within the template
+	funcMap := template.FuncMap{"env": utils.GetCachedEnv}
```

Comment:
Rather than exposing the full environment, should we simply project certain variables?  K8s has a tendency to inject a TON of environment variables, and I wonder if that's part of why you are hitting this.

e.g. if we exposed a certain fixed-set of variables as fields in `templateInput` then we could drop the whole `GetCachedEnv` thing and persist a small fixed group populated at startup?

Reply:
I expect environment variables to give flexibility to vendors to inject their specific information. For example, on GKE, we need to put project id into the trace id - in order to do that, we will need to inject a Google specific environment variable into the pods and write a template to consume that.